File: src/exports/LandRegistry.cc
#include "pland/PLand.h"

#include "exports/APIHelper.h"

#include "pland/BuildInfo.h"
#include "pland/aabb/LandAABB.h"
#include "pland/land/Land.h"
#include "pland/land/repo/LandRegistry.h"
#include "pland/utils/JsonUtil.h"

#include "mc/platform/UUID.h"

#include <algorithm>
#include <vector>

#include "ExportDef.h"


namespace ldapi {


void Export_Class_LandRegistry() {
    exportAs("LandRegistry_isOperator", [](std::string const& uuid) -> bool {
        if (!mce::UUID::canParse(uuid)) {
            return false;
        }
        return land::PLand::getInstance().getLandRegistry().isOperator(mce::UUID{uuid});
    });

    exportAs("LandRegistry_addOperator", [](std::string const& uuid) -> bool {
        if (!mce::UUID::canParse(uuid)) {
            return false;
        }
        return land::PLand::getInstance().getLandRegistry().addOperator(mce::UUID{uuid});
    });

    exportAs("LandRegistry_removeOperator", [](std::string const& uuid) -> bool {
        if (!mce::UUID::canParse(uuid)) {
            return false;
        }
        return land::PLand::getInstance().getLandRegistry().removeOperator(mce::UUID{uuid});
    });

    exportAs("LandRegistry_getOperators", []() -> std::vector<std::string> {
        auto& ops = land::PLand::getInstance().getLandRegistry().getOperators();
        return std::vector<std::string>{ops.begin(), ops.end()};
    });

    exportAs("LandRegistry_getOrCreatePlayerSettings", [](std::string const& uuid) -> std::string {
        if (!mce::UUID::canParse(uuid)) {
            return {};
        }
        try {
            // struct -> json -> std::string
            auto& settings = land::PLand::getInstance().getLandRegistry().getOrCreatePlayerSettings(mce::UUID{uuid});
            auto  j        = land::json_util::struct2json(settings);
            return j.dump();
        } catch (...) {
            return {};
        }
    });

    exportAs("LandRegistry_hasLand", [](int id) -> bool {
        return land::PLand::getInstance().getLandRegistry().hasLand(id);
    });

    exportAs(
        "LandRegistry_addOrdinaryLand",
        [](InternalLandAABB iaabb, bool is3D, std::string const& owner) -> std::string {
            if (iaabb[0].second != iaabb[1].second) {
                throw std::runtime_error("LandRegistry_addOrdinaryLand: Invalid AABB");
            }
            if (!mce::UUID::canParse(owner)) {
                throw std::runtime_error("LandRegistry_addOrdinaryLand: Invalid owner");
            }
            auto dimId = iaabb[0].second;
            auto aabb  = toCpp<land::LandAABB>(iaabb);
            aabb.fix();

            auto land     = land::Land::make(aabb, dimId, is3D, mce::UUID{owner});
            auto expected = land::PLand::getInstance().getLandRegistry().addOrdinaryLand(land);
            return asRPCResultWithValue(expected, [land]() { return land->getId(); });
        }
    );

    exportAs("LandRegistry_getLand", [](int id) -> int {
        auto land = land::PLand::getInstance().getLandRegistry().getLand(id);
        if (!land) return -1;
        return land->getId();
    });

    exportAs("LandRegistry_removeOrdinaryLand", [](int id) -> int {
        auto ptr    = land::PLand::getInstance().getLandRegistry().getLand(id);
        auto result = land::PLand::getInstance().getLandRegistry().removeOrdinaryLand(ptr);
        return asRPCResultVoid(result);
    });

    using LandList = std::vector<land::LandID>;
    exportAs("LandRegistry_getLands", []() -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry().getLands();
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands1", [](int dimid) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry().getLands(dimid);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands2", [](std::string const& uuid, bool includeShared) -> LandList {
        if (!mce::UUID::canParse(uuid)) {
            return {};
        }
        auto     lands = land::PLand::getInstance().getLandRegistry().getLands(mce::UUID{uuid}, includeShared);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands3", [](std::string const& uuid, int dimid) -> LandList {
        if (!mce::UUID::canParse(uuid)) {
            return {};
        }
        auto     lands = land::PLand::getInstance().getLandRegistry().getLands(mce::UUID{uuid}, dimid);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLands4", [](std::vector<int> lds) -> LandList {
        LandList int64List;
        int64List.reserve(lds.size());
        std::transform(lds.begin(), lds.end(), std::back_inserter(int64List), [](int i) {
            return static_cast<land::LandID>(i);
        });

        auto lands = land::PLand::getInstance().getLandRegistry().getLands(int64List);

        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });

        return result;
    });

    exportAs("LandRegistry_getPermType", [](std::string const& uuid, int landID, bool includeOperator) {
        return static_cast<int>(land::PLand::getInstance()
                                    .getLandRegistry()
                                    .getPermType(mce::UUID{uuid}, static_cast<land::LandID>(landID), includeOperator));
    });

    exportAs("LandRegistry_getLandAt", [](IntPos const& pos) -> int {
        auto land = land::PLand::getInstance().getLandRegistry().getLandAt(pos.first, pos.second);
        if (!land) return -1;
        return land->getId();
    });

    exportAs("LandRegistry_getLandAt1", [](IntPos const& pos, int radius) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry().getLandAt(pos.first, radius, pos.second);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_getLandAt2", [](IntPos const& a, IntPos const& b) -> LandList {
        auto     lands = land::PLand::getInstance().getLandRegistry().getLandAt(a.first, b.first, a.second);
        LandList result;
        result.reserve(lands.size());
        std::transform(lands.begin(), lands.end(), std::back_inserter(result), [](auto& land) {
            return land->getId();
        });
        return result;
    });

    exportAs("LandRegistry_refreshLandRange", [](int id) -> void {
        auto& inst = land::PLand::getInstance().getLandRegistry();
        auto  land = inst.getLand(id);
        if (land) inst.refreshLandRange(land);
    });

    exportAs("PLand_getVersionMeta", []() -> std::string {
        static struct {
            std::string Commit = land::BuildInfo::Commit.data();
            std::string Branch = land::BuildInfo::Branch.data();
            std::string Tag    = land::BuildInfo::Tag.data();
        } meta;
        static auto res = land::json_util::struct2json(meta).dump();
        return res;
    });
}


} // namespace ldapi

File: ts/imports/LandRegistry.ts
import {
    FfiPayload,
    Expected,
    importSymbol,
    isIntPos,
    LandID,
    LandPermType,
    UUID, InternalLandAABB, FfiProtocol, asExpected,
} from "../ImportDef.js";
import {LandAABB} from "./LandAABB.js";
import {Land} from "./Land.js";

/**
 * @warning 请不要增加、删除 key，否则会导致反射失败
 */
export type PlayerSettings = {
    /** 是否显示进入领地提示 */ showEnterLandTitle: boolean;
    /** 是否持续显示底部提示 */ showBottomContinuedTip: boolean;
};

export class LandRegistry {
    static IMPORTS = {
        LandRegistry_isOperator: importSymbol("LandRegistry_isOperator"),
        LandRegistry_addOperator: importSymbol("LandRegistry_addOperator"),
        LandRegistry_removeOperator: importSymbol("LandRegistry_removeOperator"),
        LandRegistry_getOperators: importSymbol("LandRegistry_getOperators") as () => UUID[],
        LandRegistry_getOrCreatePlayerSettings: importSymbol(
            "LandRegistry_getOrCreatePlayerSettings",
        ),
        LandRegistry_getLand: importSymbol("LandRegistry_getLand"),
        LandRegistry_hasLand: importSymbol("LandRegistry_hasLand"),
        LandRegistry_getLands: importSymbol("LandRegistry_getLands"),
        LandRegistry_getLands1: importSymbol("LandRegistry_getLands1"),
        LandRegistry_getLands2: importSymbol("LandRegistry_getLands2"),
        LandRegistry_getLands3: importSymbol("LandRegistry_getLands3"),
        LandRegistry_getLands4: importSymbol("LandRegistry_getLands4"),
        LandRegistry_getPermType: importSymbol("LandRegistry_getPermType"),
        LandRegistry_getLandAt: importSymbol("LandRegistry_getLandAt"),
        LandRegistry_getLandAt1: importSymbol("LandRegistry_getLandAt1"),
        LandRegistry_getLandAt2: importSymbol("LandRegistry_getLandAt2"),
        LandRegistry_refreshLandRange: importSymbol(
            "LandRegistry_refreshLandRange",
        ),
        PLand_getVersionMeta: importSymbol("PLand_getVersionMeta"),
        LandRegistry_removeOrdinaryLand: importSymbol(
            "LandRegistry_removeOrdinaryLand",
        ) as (id: LandID) => FfiProtocol,
        LandRegistry_addOrdinaryLand: importSymbol("LandRegistry_addOrdinaryLand") as (aabb: InternalLandAABB, is3D: boolean, owner: UUID) => FfiProtocol,

        LandRegistry_createSnapshot: importSymbol("LandRegistry_createSnapshot") as (dirName?: string) => void,
    };

    constructor() {
        throw new Error("LandRegistry is a static class");
    }

    /**
     * 创建数据库快照
     * @param dirName 快照文件夹名称，如果为空，则使用当前时间戳
     * @note 创建的快照会被写入磁盘, snapshots/<dirName ?? timestamp>
     * @note 此任务为异步任务，如果任务未完成，文件夹下会存在 .incomplete 文件
     * @version v0.19.x
     */
    static createSnapshot(dirName?: string): void {
        // LegacyRemoteCall 不支持 optional，这里采用空字符串 = null、undefined
        return LandRegistry.IMPORTS.LandRegistry_createSnapshot(dirName ?? "");
    }

    static isOperator(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_isOperator(uuid);
    }

    static addOperator(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_addOperator(uuid);
    }

    static removeOperator(uuid: string): boolean {
        return LandRegistry.IMPORTS.LandRegistry_removeOperator(uuid);
    }

    static getOperators() {
        return LandRegistry.IMPORTS.LandRegistry_getOperators();
    }

    static getOrCreatePlayerSettings(uuid: string): PlayerSettings | null {
        const jsonStr =
            LandRegistry.IMPORTS.LandRegistry_getOrCreatePlayerSettings(uuid);
        try {
            return JSON.parse(jsonStr);
        } catch (e) {
            logger.error(`Failed to parse player settings: ${jsonStr}`);
            return null;
        }
    }

    static hasLand(id: LandID): boolean {
        return LandRegistry.IMPORTS.LandRegistry_hasLand(id);
    }

    /**
     * 创建并添加一个普通领地
     * @param aabb 领地范围
     * @param is3D 是否是3D领地
     * @param owner 领地拥有者(UUID)
     * @returns Expected<Land>
     */
    static createAndAddOrdinaryLand(
        aabb: LandAABB,
        is3D: boolean,
        owner: UUID,
    ): Expected<Land> {
        if (aabb.min.dimid != aabb.max.dimid) {
            throw new Error("AABB min and max must be in the same dimension");
        }
        const protocol = LandRegistry.IMPORTS.LandRegistry_addOrdinaryLand(
            [aabb.min, aabb.max],
            is3D,
            owner,
        );
        return asExpected<LandID>(protocol).map(id => new Land(id));
    }

    /**
     * @brief 移除普通领地
     */
    static removeOrdinaryLand(
        land: Land | LandID,
    ): Expected<void> {
        const id =
            typeof land === "number"
                ? land
                :
                (land as Land).mLandId;
        const protocol = LandRegistry.IMPORTS.LandRegistry_removeOrdinaryLand(id);
        return asExpected<void>(protocol);
    }

    static getLand(landID: LandID): Land | null {
        const id = LandRegistry.IMPORTS.LandRegistry_getLand(landID);
        if (id === -1) {
            return null;
        }
        return new Land(id);
    }

    /**
     * 获取所有领地
     */
    static getLands(): Land[];
    /**
     * 批量获取领地
     * @param ids id 列表
     */
    static getLands(ids: LandID[]): Land[];
    /**
     * 获取指定维度的所有领地
     * @param dimid 维度 id
     */
    static getLands(dimid: number): Land[];
    /**
     * 获取玩家所有的领地
     * @param uuid 玩家 uuid
     * @param includeShared 是否包含其它玩家共享的领地(默认false)
     */
    static getLands(uuid: UUID, includeShared: boolean): Land[];
    /**
     * 获取玩家在指定维度所有的领地
     * @param uuid 玩家 uuid
     * @param dimid 维度 id
     */
    static getLands(uuid: UUID, dimid: number): Land[];

    static getLands(...args: any[]): Land[] {
        switch (args.length) {
            case 0: {
                // getLands() const;
                return LandRegistry.IMPORTS.LandRegistry_getLands().map(
                    (id: LandID) => new Land(id),
                );
            }
            case 1: {
                if (Array.isArray(args[0])) {
                    // getLands(std::vector<LandID> const& ids) const;
                    return LandRegistry.IMPORTS.LandRegistry_getLands4(
                        args[0],
                    ).map((id: LandID) => new Land(id));
                } else if (typeof args[0] === "number") {
                    // getLands(LandDimid dimid) const;
                    return LandRegistry.IMPORTS.LandRegistry_getLands1(
                        args[0],
                    ).map((id: LandID) => new Land(id));
                }
            }
            case 2: {
                const arg0 = args[0];
                if (typeof arg0 === "string") {
                    if (typeof args[1] === "boolean" || args[1] === undefined) {
                        // getLands(UUID const& uuid, bool includeShared = false) const;
                        const includeShared = args[1] ?? false;
                        return LandRegistry.IMPORTS.LandRegistry_getLands2(
                            arg0,
                            includeShared,
                        ).map((id: LandID) => new Land(id));
                    } else {
                        // getLands(UUID const& uuid, LandDimid dimid) const;
                        return LandRegistry.IMPORTS.LandRegistry_getLands3(
                            arg0,
                            args[1],
                        ).map((id: LandID) => new Land(id));
                    }
                }
            }
        }
        throw new Error("Invalid arguments");
    }

    /**
     * 查询指定坐标的领地
     * @param pos 坐标
     */
    static getLandAt(pos: IntPos): Land | null;
    /**
     * 查询指定坐标半径内的领地
     * @param pos 坐标
     * @param radius 半径
     */
    static getLandAt(pos: IntPos, radius: number): Land[];
    /**
     * 查询 AABB 区域内的领地
     * @param pos1 min 坐标
     * @param pos2 max 坐标
     */
    static getLandAt(pos1: IntPos, pos2: IntPos): Land[];

    static getLandAt(
        pos1: IntPos,
        radius_or_pos2?: number | IntPos,
    ): Land | null | Land[] {
        if (!isIntPos(pos1)) {
            throw new Error("Invalid arguments");
        }

        if (radius_or_pos2 === undefined) {
            // getLandAt(BlockPos const& pos, LandDimid dimid) const;
            const id = LandRegistry.IMPORTS.LandRegistry_getLandAt(pos1);
            if (id === -1) {
                return null;
            }
            return new Land(id);
        } else if (typeof radius_or_pos2 === "number") {
            // getLandAt(BlockPos const& center, int radius, LandDimid dimid) const;
            return LandRegistry.IMPORTS.LandRegistry_getLandAt1(
                pos1,
                radius_or_pos2,
            ).map((id: LandID) => new Land(id));
        } else if (isIntPos(radius_or_pos2)) {
            // getLandAt(BlockPos const& pos1, BlockPos const& pos2, LandDimid dimid) const;
            if (pos1.dimid != (radius_or_pos2 as IntPos).dimid) {
                throw new Error("Invalid arguments");
            }
            return LandRegistry.IMPORTS.LandRegistry_getLandAt2(
                pos1,
                radius_or_pos2,
            ).map((id: LandID) => new Land(id));
        } else {
            throw new Error("Invalid arguments");
        }
    }

    static getPermType(
        uuid: UUID,
        landID = 0,
        includeOperator = true,
    ): LandPermType {
        return LandRegistry.IMPORTS.LandRegistry_getPermType(
            uuid,
            landID,
            includeOperator,
        );
    }

    static refreshLandRange(land: Land): void {
        // @ts-ignore
        LandRegistry.IMPORTS.LandRegistry_refreshLandRange(land.unique_id);
    }

    static getVersionMeta(): {
        Commit: string;
        Branch: string;
        Tag: string;
    } {
        return JSON.parse(LandRegistry.IMPORTS.PLand_getVersionMeta());
    }
}

Object.freeze(LandRegistry.IMPORTS);
